File: backend/app/routers/notifications.py
from fastapi import APIRouter, Depends, status
from motor.motor_asyncio import AsyncIOMotorDatabase

from app.db.mongo import get_db
from app.core.dependencies import get_current_user
from app.models.notification import NotificationResponse
from app.repositories.notification_repo import NotificationRepository
from app.services.notification_service import NotificationService

router = APIRouter(
    prefix="/notifications",
    tags=["Notifications"],
)


def get_notification_service(
    db: AsyncIOMotorDatabase = Depends(get_db),
) -> NotificationService:
    repo = NotificationRepository(db)
    return NotificationService(repo)


@router.get("/", response_model=list[NotificationResponse])
async def my_notifications(
    current_user: dict = Depends(get_current_user),
    service: NotificationService = Depends(get_notification_service),
):
    notifications = await service.my_notifications(current_user)

    return [
        NotificationResponse(
            id=str(n["_id"]),
            user_id=str(n["user_id"]),
            type=n["type"],
            message=n["message"],
            task_id=str(n["task_id"]) if n.get("task_id") else None,
            is_read=n["is_read"],
            created_at=n["created_at"],
        )
        for n in notifications
    ]


@router.post(
    "/{notification_id}/read",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def mark_notification_read(
    notification_id: str,
    current_user: dict = Depends(get_current_user),
    service: NotificationService = Depends(get_notification_service),
):
    await service.mark_as_read(notification_id)

File: frontend/src/lib/notifications.ts
import { api } from './api'

export interface Notification {
  id: string
  user_id: string
  type: string
  message: string
  task_id?: string | null
  is_read: boolean
  created_at: string
}

export async function getNotifications(): Promise<Notification[]> {
  const { data } = await api.get('/notifications/')
  return data
}

export async function markNotificationRead(notificationId: string): Promise<void> {
  await api.post(`/notifications/${notificationId}/read`)
}

export async function markAllNotificationsRead(notifications: Notification[]): Promise<void> {
  const unread = notifications.filter(n => !n.is_read)
  await Promise.all(unread.map(n => markNotificationRead(n.id)))
}
